# MATH&ML-2 Линейная алгебра в констекте линейных методов. Часть2
###  Содержание <a class="anchor" id=0></a>

- [2. Неоднородные СЛАУ](#2)
- [3. Линейная регрессия МНК](#3)
- [4. Стандартизация векторов и матрица корреляции](#4)
- [5. Практика. Лин.регрессия МНК](#5)
- [6. Полиноминальная регрессия](#6)
- [7. Регуляция](#7)
- [8. Практика. Полиноминальная регрессия и регуляция](#8)
- [9. Итоги](#9)

# 2. Неоднородные СЛАУ <a class="anchor" id=2></a>

[к содержанию](#0)


>Совокупность уравнений первой степени, в которых каждая переменная и коэффициенты в ней являются вещественными числами, называется **системой линейных алгебраических уравнений** (`СЛАУ`) и в общем случае записывается как:
>
>$\left\{ \begin{array}{c} a_{11}x_1+a_{12}x_2+\dots +a_{1m}x_m=b_1 \\ a_{21}x_1+a_{22}x_2+\dots +a_{2m}x_m=b_2 \\ \dots \\ a_{n1}x_1+a_{n2}x_2+\dots +a_{nm}x_m=b_n \end{array} \right.\ (1),$
>
>где
>
>* $n$— количество уравнений;
>
>* $m$ — количество переменных;
>
>* $x_i$ — неизвестные переменные системы;
>
>* $a_{ij}$ — коэффициенты системы;
>
>* $b_i$ — свободные члены системы.
>
>
>
>СЛАУ (1) называется **однородной**, если все свободные члены системы равны 0 $b_1=b_2=⋯=b_n=0$:
>
>$\textrm{С}\textrm{Л}\textrm{А}\textrm{У}-\textrm{о}\textrm{д}\textrm{н}\textrm{о}\textrm{р}\textrm{о}\textrm{д}\textrm{н}\textrm{а}\textrm{я},\ \textrm{е}\textrm{с}\textrm{л}\textrm{и}\ \forall b_i=0$
>
>
>
>СЛАУ (1) называется **неоднородной**, если хотя бы один из свободных членов системы отличен от 0:
>
>$\textrm{С}\textrm{Л}\textrm{А}\textrm{У}- \textrm{н}\textrm{е}\textrm{о}\textrm{д}\textrm{н}\textrm{о}\textrm{р}\textrm{о}\textrm{д}\textrm{н}\textrm{а}\textrm{я},\ \textrm{е}\textrm{с}\textrm{л}\textrm{и}\ \exists b_i\neq 0$
>
>
>
>**Решением** СЛАУ (1) называется такой набор значений неизвестных переменных $x_1,x_2,…,x_n$ при котором каждое уравнение системы превращается в равенство.
>
>
>
>СЛАУ (1) называется **определённой**, если она имеет только одно решение, и **неопределённой**, если возможно больше одного решения.

Вспомним, что СЛАУ можно записать в матричном виде:

$A\overrightarrow{x}=\overrightarrow{b}$

$\left( \begin{array}{cccc} a_{11} & a_{12} & \dots & a_{1m} \\ a_{21} & a_{22} & \dots & a_{2m} \\ \dots & \dots & \dots & \dots \\ a_{n1} & a_{n2} & \dots & a_{nm} \end{array} \right) \cdot \left( \begin{array}{c} x_1 \\ x_2 \\ \dots \\ x_m \end{array} \right)=\left( \begin{array}{c} b_1 \\ b_2 \\ \dots \\ b_n \end{array} \right)$

где $A$ — матрица системы, $\overrightarrow{x}$ — вектор неизвестных коэффициентов, а $b$ — вектор свободных коэффициентов. 

Давайте введём новое для нас определение.

>**Расширенной матрицей системы $(A|b)$ неоднородных СЛАУ** называется матрица, составленная из исходной матрицы и вектора свободных коэффициентов (записывается через вертикальную черту):
>
>$(A \mid \vec{b})=\left(\begin{array}{cccc|c} a_{11} & a_{12} & \ldots & a_{1 m} & b_{1} \\ a_{21} & a_{22} & \ldots & a_{2 m} & b_{2} \\ \ldots & \ldots & \ldots & \ldots & \ldots \\ a_{n 1} & a_{n 2} & \ldots & a_{n m} & b_{n} \end{array}\right)$

Расширенная матрица системы — это обычная матрица. Черта, отделяющая коэффициенты $a_{ij}$ от свободных членов $b_i$ — чисто символическая. 

Над расширенной матрицей неоднородной СЛАУ можно производить те же самые действия, что и над обычной, а именно:

* складывать/вычитать между собой строки/столбцы матрицы;
* умножать строки/столбцы на константу;
* менять строки/столбцы местами.



Приведём пример расширенной матрицы системы. Пусть исходная система будет следующей:

$\left\{\begin{array}{c} w_{1}+w_{2}=1 \\ w_{1}+2 w_{2}=2 \end{array}\right.$

Запишем её в матричном виде:

$\left(\begin{array}{cc} 1 & 1 \\ 1 & 2  \end{array} \right) \cdot \left(\begin{array}{c} w_1 \\ w_2  \end{array} \right) = \left(\begin{array}{c} 1 \\ 2  \end{array} \right)$

Тогда расширенная матрица системы будет иметь вид:

$(A \mid b)=\left(\begin{array}{cc|c} 1 & 2 & 1 \\ 1 & 2 & 2 \\ \end{array}\right)$

***

Существует три случая при решении неоднородных СЛАУ:

* **«Идеальная пара»**

Это так называемые определённые системы линейных уравнений, имеющие **единственные решения**.

* **«В активном поиске»**

Неопределённые системы, имеющие **бесконечно много решений**.

* **«Всё сложно»**

Это самый интересный для нас случай — переопределённые системы, которые **не имеют точных решений**.

>**Примечание**. В данной классификации неоднородных СЛАУ допущено упрощение в терминологии. На самом деле неопределённые системы — это те, в которых независимых уравнений меньше, чем неизвестных. Они могут иметь бесконечно много решений (быть совместными) или ни одного решения (быть несовместными, если уравнения противоречат друг другу).
>
>На практике, например в обучении регрессий, этот случай практически не встречается.
>
>Что касается переопределённых систем, то в них, помимо несовместности (отсутствия решений), количество независимых уравнений превышает количество неизвестных — это тот самый случай, что мы видим в регрессионном анализе.

## СЛУЧАЙ «ИДЕАЛЬНАЯ ПАРА»

>Самый простой случай решения неоднородной СЛАУ — когда система **имеет единственное решение**. Такие системы называются **совместными**.

На вопрос о том, когда СЛАУ является совместной, отвечает главная теорема СЛАУ — теорема **Кронекера — Капелли** (также её называют **критерием совместности системы**).

***

**Теорема Кронекера — Капелли:**

Неоднородная система линейный алгебраических уравнений $A\overrightarrow{w} = \overrightarrow{b}$ является совместной тогда и только тогда, когда ранг матрицы системы $A$ равен рангу расширенной матрицы системы $(A|b)$ и равен количеству независимых переменных $m$:

$rk(A) = rk(A|\overrightarrow{b}) = m \leftrightarrow \exists ! \overrightarrow{w} = (w_{1}, w_{2}, \ldots w_m)^T$

Причём решение системы будет равно:

$\overrightarrow{w} = A^{-1} \overrightarrow{b}$

***



Пример №1

$\left\{\begin{array}{c} w_{1}+w_{2}=1 \\ w_{1}+2 w_{2}=2 \end{array}\right.$

где  $w_1$ и $w_2$ — неизвестные переменные.

При решении системы «в лоб» получим:

$\left\{\begin{array}{c} w_{1}+w_{2}=1 \\ w_{1}+2 w_{2}=2 \end{array}\right. \Rightarrow \left\{\begin{array}{c} w_{1}+w_{2}=1 \\ w_{2}=2 \end{array}\right. \Rightarrow \left\{\begin{array}{c} w_{1}=0 \\ w_{2}=0 \end{array}\right.$

Интерпретация:

$\left( \begin{array}{c} 1 \\ 2 \end{array}\right) = 0 \cdot \left( \begin{array}{c} 1 \\ 1 \end{array}\right) + 1 \cdot \left( \begin{array}{c} 1 \\ 2 \end{array}\right)$

На языке линейной алгебры это означает что вектор $(1, 2)^T$ линейно выражается через векторы коэффициентов системы $(1, 1)^T$ и $(1, 2)^T$.

В матричном виде система запишется, как:

$A\overrightarrow{w}=\overrightarrow{b} \text{где}A = \left( \begin{array}{cc} 1 & 1 \\ 1 & 2 \end{array}\right), \overrightarrow{w}=\left( \begin{array}{c} w_1 \\ w_2 \end{array}\right), \overrightarrow{b}=\left( \begin{array}{c} b_1 \\ b_2 \end{array}\right) $ 

Преобразование уравнений будем таким же, как и при преобразовании расширенной матрицы системы $(A|b)$, вычитая сначала первую строку из второй, а затем — результат из первой, получим то же решение, что и решение «в лоб».

$(A|\overrightarrow{b})=\left(\begin{array}{cc|c} 1 & 1 & 1  \\ 1 & 2 & 2 \end{array} \right) \Rightarrow \left(\begin{array}{cc|c} 1 & 1 & 1  \\ 0 & 1 & 1 \end{array} \right) \Rightarrow \left(\begin{array}{cc|c} 1 & 0 & 0  \\ 0 & 1 & 1 \end{array} \right) \Rightarrow \left\{\begin{array}{c} w_1=0  \\ w_2=0 \end{array} \right.$

Других решений у системы нет. 

Посмотрим на ранги матрицы $А$ и расширенной матрицы $(A|b)$ (количество ступеней в ступенчатых матрицах):

$rk(A)=\left(\begin{array}{cc} 1 & 0 \\ 0 & 1\end{array}\right)=2$

$rk(A|b)=\left(\begin{array}{ccc} 1 & 0 & 0 \\ 0 & 1 & 1\end{array}\right)=2$

$rk(A)=rk(A|b)$

Они совпадают и равны количеству неизвестных, а это и гарантирует существование и **единственность решения**. То есть в общем случае, чтобы узнать, сколько решений существует у системы, её необязательно было бы решать — достаточно было бы найти ранги матриц $rk(A)$ и $rk(A|b)$ .

***
Тут возникает вопрос: «Можно ли найти решение одной формулой?»

Для удобства перепишем систему без стрелок:

$Aw=b$

Так как матрица квадратная и невырожденная, у неё обязательно есть обратная матрица.

Умножим на $A^{-1}$ слева обе части уравнения. Стоит напомнить, что произведение матриц **не перестановочно**, поэтому есть разница, с какой стороны умножать.

$A^{-1} \cdot Aw = A^{-1}\cdot b$

$w=A^{-1}\cdot b$

>**Важно**! Отсюда явно видны ограничения этого метода: его можно применять только **для квадратных невырожденных матриц** (тех, у которых определитель не равен 0).

Убедимся в правильности формулы. Найдём произведение матрицы $A^{-1}$ и вектора-столбца $b$:

$A^{-1}\cdot b=\left( \begin{array}{cc} 1 & 1 \\ 1 & 2 \end{array}\right)^{-1} \cdot \left(\begin{array}{c} 1 \\ 2 \end{array}\right)=\left( \begin{array}{cc} 2 & -1 \\ -1 & 1 \end{array}\right) \cdot \left(\begin{array}{c} 1 \\ 2 \end{array}\right) = \left(\begin{array}{c} 0 \\ 1 \end{array}\right)$ 

***

**Резюмируем ↓**

У нас есть квадратная система с $m$ неизвестных. Если ранг матрицы коэффициентов $A$ **равен** рангу расширенной матрицы $(A|b)$ и **равен** количеству переменных $(rk(A)=rk(\overrightarrow{b}))=m$, то в системе будет ровно столько независимых уравнений, сколько и неизвестных $m$, а значит будет **единственное** решение.

Вектор свободных коэффициентов $b$ при этом линейно независим со столбцами матрицы $A$, его разложение по столбцам $A$ **единственно**.

## СЛУЧАЙ «В АКТИВНОМ ПОИСКЕ»

?А что, если система **не удовлетворяет теореме Кронекера — Капелли**? То есть ранг матрицы системы равен расширенному рангу матрицы, но не равен количеству неизвестных. Неужели тогда система нерешаема?

На этот вопрос отвечает первое следствие из теоремы ↓

**Следствие №1** из теоремы Кронекера — Капелли:

Если ранг матрицы системы $A$ равен рангу расширенной матрицы системы $(A|b)$, **но меньше**, чем количество неизвестных $m$, то система имеет бесконечное множество решений:

$rk(A) = rk(A | \vec{b}) < m  \leftrightarrow  \infty \ решений$

Решим систему уравнений:

$w_1+w_2+w_3=10$

Да, уравнение одно, но формально оно является неоднородной СЛАУ.

Итак, мы имеем одно уравнение на три неизвестных, значит две координаты из трёх вектора $w$ мы можем задать как угодно. Например, зададим вторую и третью как $\alpha$ и $\beta$. Тогда первая будет равна $10-\alpha-\beta$.

$w=\left( \begin{array}{c} {10-\alpha-\beta} \\ \alpha \\ \beta \end{array} \right)$ где $\alpha, \beta \in \mathbb{R}$

Вместо переменных $\alpha$ и $\beta$ мы можем подставлять любые числа и всегда будем получать равенство. 

Составим расширенную матрицу:

$(A|b)=(\begin{array}{} 1 & 1 & 1|10 \end{array})$

Её ранг, как и ранг $A$, равен 1, что меньше числа неизвестных $m=3$:

$rk(A) = rk(A | \vec{b}) = 1 < 3$

Такая ситуация, по следствию из теоремы Кронекера — Капелли, говорит о существовании и не единственности решения, то есть решений **бесконечно много**.

***

**Резюмируем ↓
**
Если ранги матриц $A$ и $(A|b)$ всё ещё совпадают, но уже меньше количества неизвестных $(rk(A) = rk(A | \vec{b}) < m)$, значит, уравнений не хватает для того, чтобы определить систему полностью, и решений будет бесконечно много.

На языке линейной алгебры это значит, что вектор $\overrightarrow{b}$ линейно зависим со столбцами матрицы $A$, но также и сами столбцы зависимы между собой, поэтому равнозначного разложения не получится, т. е. таких разложений может быть сколько угодно.

## СЛУЧАЙ «ВСЁ СЛОЖНО»

А теперь посмотрим на самый интересный для нас случай. Его формально регламентирует второе следствие из теоремы Кронекера — Капелли.

**Следствие №2 из теоремы Кронекера — Капелли**:

Если ранг матрицы системы $A$ меньше, чем ранг расширенной матрицы системы $(A|b)$, то система несовместна, то есть не имеет точных решений:

$rk(A)  < rk(A | \vec{b})  \leftrightarrow  \nexists \ решений$

Решим систему уравнений:

$\left\{\begin{array}{c} w_1+w_2=1 \\ w_1+2 w_2=2 \\ w_1+w_2=12 \end{array}\right.$

Посмотрим на первое и третье уравнение — очевидно, что такая система не имеет решений, так как данные уравнения противоречат друг другу.

Но давайте обоснуем это математически. Для этого запишем расширенную матрицу системы:

$(A|b)=\left(\begin{array}{cc|c} 1 & 1 & 1 \\ 1 & 2 & 2 \\ 1 & 1 & 12 \end{array}\right)$

Посчитаем ранги матриц $A$ и $(A|b)$:

$rk(A)=rk\left(\begin{array}{cc} 1 & 1  \\ 1 & 2 \\ 1 & 1 \end{array}\right) \Rightarrow I-III \Rightarrow rk\left(\begin{array}{cc} 1 & 1  \\ 1 & 2 \end{array}\right) \Rightarrow II-I \Rightarrow rk\left(\begin{array}{cc} 1 & 1  \\ 0 & 1 \end{array}\right)=2$

$rk(A|b)=rk\left(\begin{array}{ccc} 1 & 1 & 1 \\ 1 & 2 & 2 \\ 1 & 1 & 12 \end{array}\right) \Rightarrow III-I, II-I \Rightarrow rk\left(\begin{array}{cc} 1 & 1 & 1 \\ 0 & 1 & 1 \\ 0 & 0 & 12\end{array}\right)=3$

Итак, $rk(A)=2$, в то время как $rk(A|b)=3$. Это и есть критерий **переопределённости системы уравнений**: ранг матрицы системы меньше ранга расширенной матрицы системы.

Получается, что идеальное решение найти нельзя, но чуть позже мы увидим, что такие системы возникают в задачах регрессии практически всегда, а значит нам всё-таки хотелось бы каким-то образом её решать. Можно попробовать найти приблизительное решение — вопрос лишь в том, какое из всех этих решений лучшее.

Найдем наилучшее приближение для $w_1$, $w_2$, если:

$\left\{\begin{array}{c} w_1+w_2=1 \\ w_1+2 w_2=2 \\ w_1+w_2=12 \end{array}\right. \Rightarrow \left(\begin{array}{cc} 1 & 1  \\ 1 & 2 \\ 1 & 1 \end{array}\right)\cdot \left(\begin{array}{c} w  \\ w \end{array}\right)=\left(\begin{array}{c} 1 \\ 2 \\ 12 \end{array}\right)$

Обозначим приближённое решение как $\hat{w}$. Приближением для вектора $b$ будет $\hat{b}=A \hat{w}$. Также введём некоторый вектор ошибок $e=b-\hat{b}=b-A \hat{w}$.

>**Примечание**. Здесь мы снова опустили стрелки у векторов $b$, $\hat{b}$ и $\hat{w}$ для наглядности.

Например, если мы возьмём в качестве вектора $\hat{w}$ вектор $\hat{w_1}=(1,1)^T$, то получим:

$\hat{b}=A\hat{w_1}=\left(\begin{array}{cc} 1 & 1 \\ 1 & 2 \\ 1 & 1\end{array}\right)\cdot\left(\begin{array}{c} 1 \\ 1  \end{array}\right)=\left(\begin{array}{c} 2 \\ 3 \\ 2 \end{array}\right)$

$e_1=b-A\hat{w_1}=\left(\begin{array}{c} 1 \\ 2 \\ 12\end{array}\right) - \left(\begin{array}{c} 2 \\ 3 \\ 2\end{array}\right)=\left(\begin{array}{c} -1 \\ -1 \\ 10 \end{array}\right)$

Теперь возьмём в качестве вектора $\hat{w_2}=(4, -1)^T$, получим:

$\hat{b}=A\hat{w_2}=\left(\begin{array}{cc} 1 & 1 \\ 1 & 2 \\ 1 & 1\end{array}\right)\cdot\left(\begin{array}{c} 4 \\ -1  \end{array}\right)=\left(\begin{array}{c} 3 \\ 2 \\ 3 \end{array}\right)$

$e_2=b-A\hat{w_2}=\left(\begin{array}{c} 1 \\ 2 \\ 12\end{array}\right) - \left(\begin{array}{c} 3 \\ 2 \\ 3\end{array}\right)=\left(\begin{array}{c} -2 \\ -1 \\ 9 \end{array}\right)$

>Конечно, нам хотелось бы, чтобы ошибка была поменьше. Но какая из них поменьше? Векторы сами по себе сравнить нельзя, **но зато можно сравнить их длины**.

$\left\|e_1 \right\| = \sqrt{(-1)^2 + (-1)^2 + (10)^2} = \sqrt{102}$

$\left\|e_2 \right\| = \sqrt{(-2)^2 + 0^2 + 9^2} = \sqrt{85}$

>Видно, что вторая ошибка всё-таки меньше, соответственно, приближение лучше. Но в таком случае из всех приближений нам нужно выбрать то, у которого длина вектора ошибок минимальна, если, конечно, это возможно.
>
>$||e||\rightarrow min$
>

***

>**Примечание**. Проблема поиска оптимальных приближённых решений неоднородных переопределённых СЛАУ стояла у математиков вплоть до XIX века. До этого времени математики использовали частные решения, зависящие от вида уравнений и размерности. Впервые данную задачу для общего случая решил Гаусс, опубликовав метод решения этой задачи, который впоследствии будет назван методом наименьших квадратов (МНК). В дальнейшем Лаплас прибавил к данному методу теорию вероятности и доказал оптимальность МНК-оценок с точки зрения статистики.

Сейчас мы почувствуем себя настоящими математиками и попробуем решить эту задачу самостоятельно с помощью простой геометрии и знакомых нам операций над матрицами.

Вспомним, что на языке линейной алгебры неразрешимость системы

$\left(\begin{array}{cc} 1 & 1  \\ 1 & 2 \\ 1 & 1 \end{array}\right)\cdot \left(\begin{array}{c} w  \\ w \end{array}\right)=\left(\begin{array}{c} 1 \\ 2 \\ 12 \end{array}\right)$

означает, что попытка выразить вектор $(1,2,12)^T$ через $(1,1,1)^T$ и $(1,2,1)^T$ не будет успешной, так как они **линейно независимы**.

Геометрически это означает, что вектор свободных коэффициентов $\textcolor{brown}{b}$ (коричневый) не лежит в одной плоскости со столбцами матрицы $\textcolor{blue}{A}$ (синие векторы).

<img src=m2_img1.png>

Идея состояла в том, что наилучшим приближением для коричневого вектора будет ортогональная проекция на синюю плоскость — $\textcolor{cyan}{голубой}$ вектор. Так происходит потому, что наименьший по длине вектор ошибок — $\textcolor{red}{красный}$ — должен быть перпендикулярен к синей плоскости:

$e=b-\hat{b}$

В прошлом модуле мы производили расчёты интуитивно, а теперь настала пора вывести формулу.

Давайте умножим наши уравнения слева на $A^T$:

$A^T\cdot\textcolor{cyan}{A\hat{w}}=A^T\cdot\textcolor{brown}{b}$

Идея заключается в следующем: справа мы найдём скалярное произведение столбцов матрицы $A$ на вектор $b$, а слева — произведение столбцов $A$ на приближённый вектор $\hat{b}$ (по сути, на голубую проекцию).

Упростим уравнение, перемножив всё, что содержит только числа. В левой части умножим $A^T$ на $A$, в правой — умножим $A^T$ на $b$. Тогда слева получим матрицу 2×2 — это не что иное, как матрица Грама столбцов $A$.

>Столбцы $A$ линейно независимы, а это значит, что, по свойству матрицы Грама, $A^T\cdot A$  — невырожденная квадратная матрица (её определитель не равен нулю, и для неё существует обратная матрица). Получившаяся система — один в один случай «идеальная пара» (ранг матрицы, как и ранг расширенной матрицы, равен 2, в чём несложно убедиться), а это значит, что теперь мы можем её решить.

$\left(\begin{array}{} 3 & 4 \\ 4 & 6 \end{array}\right)\cdot\left(\begin{array}{} \hat{w} \\ \hat{w} \end{array}\right) = \left(\begin{array}{} 15 \\ 17 \end{array}\right)$

$\left(\begin{array}{} 3 & 4 \\ 4 & 6 \end{array}\right)$ - матрица Грамма столбцов $A$

$A^T\cdot A=Gram(\left(\begin{array}{} 1 \\ 1 \\ 1 \end{array}\right),\left(\begin{array}{} 1 \\ 2 \\ 1 \end{array}\right))$

Но ведь мы не могли решить изначальную задачу, так как она была переопределена, а эту — можем. **Как так получилось**?

Мы потребовали, чтобы у приближения $\hat{b}$ были с векторами $(1,1,1)^T$ и $(1,2,1)^T$ такие же скалярные произведения, как у $b$. Это и означает что $\hat{b}$ — ортогональная проекция на нашу синюю плоскость, в которой лежат столбцы матрицы $A$, и в этой плоскости мы можем найти коэффициенты.

Мы с вами отлично умеем решать системы типа «Идеальная пара». Для этого нам нужно найти обратную матрицу $(A^T\cdot A)^{-1}$ и умножить на неё слева всё уравнение. Так мы и получим наше приближение:

$(A^TA)\cdot\hat{w}=A^Tb$

Находим определитель матрицы $(A^TA)$: 

$\mathbb{det}(A^TA) = 3\cdot6-4\cdot4=2$

Находим обратную матрицу $(A^TA)^{-1}$:

$(A^TA)^{-1}=\left(\begin{array}{cc} 3 & 4 \\ 4 & 6\end{array}\right)^{-1}=\frac{1}{2}\left(\begin{array}{cc} 6 & -4 \\ -4 & 3\end{array}\right)=\left(\begin{array}{cc} 3 & -2 \\ -2 & 1.5\end{array}\right)$

Умножаем всё уравнение на обратную матрицу слева:

$(A^TA)^{-1}\cdot(A^TA)\cdot\hat{w}=(A^TA)^{-1}\cdot A^T\cdot b$

$\hat{w}=(A^TA)^{-1}\cdot A^T\cdot b$

И, наконец, вот он — долгожданный приближённый вектор $\hat{w}$:

$\hat{w}=\left(\begin{array}{cc} 3 & -2 \\ -2 & 1.5\end{array}\right)\cdot\left(\begin{array}{} 15 \\ 17\end{array}\right)=\left(\begin{array}{} 11 \\ -4.5\end{array}\right)$

***

⭐ **Пришло время открытий!**

Только что мы геометрическим образом вывели формулу оценки решения методом наименьших квадратов (`МНК` или `OLS`, Ordinary Least Squares).

>**Примечание**. Стоит отметить, что полученная матричная формула не зависит от размерностей и конкретных значений, а значит применима не только в нашем локальном случае, но и в общем.

Нам осталось выполнить проверку полученных результатов, чтобы убедиться в верности решения.

Вычислим голубой вектор $\hat{b}$. Для этого возьмём линейную комбинацию столбцов матрицы $А$ с найденными нами коэффициентами $\hat{w_1}$ и $\hat{w_2}$

$\hat{b}=A\hat{w}=\hat{w_1}\cdot\left(\begin{array}{} 1\\1\\1\end{array}\right)+\hat{w_2}\cdot\left(\begin{array}{} 1\\2\\1\end{array}\right)=11\cdot\left(\begin{array}{} 1\\1\\1\end{array}\right)-4.5\cdot\left(\begin{array}{} 1\\2\\1\end{array}\right)=\left(\begin{array}{} 6.5\\2\\6.5\end{array}\right)$

Вычислим вектор ошибок $e$:

$e=b-\hat{b}=b-A\hat{w}=\left(\begin{array}{} 1\\2\\12\end{array}\right)-\left(\begin{array}{} 6.5\\2\\6.5\end{array}\right)=\left(\begin{array}{} -5.5\\0\\5.5\end{array}\right)$

Убедимся, что данный вектор действительно ортогонален столбцам матрицы $А$. Для этого найдём их скалярные произведения:

$(e,A_1)=e^T\cdot A_1=(-5.5, 0, 5.5) \cdot \left(\begin{array}{} 1\\1\\1\end{array}\right) = 0$

$(e,A_2)=e^T\cdot A_2=(-5.5, 0, 5.5) \cdot \left(\begin{array}{} 1\\2\\1\end{array}\right) = 0$

Скалярные произведения равны 0, а это означает, что вектор ошибок $\textcolor{red}e$ действительно ортогонален всей синей плоскости, а голубой вектор $\textcolor{cyan}{\hat{b}}$ приближённого значения является ортогональной проекцией коричневого вектора $\textcolor{brown}b$.

>**Примечание**. Прежде чем перейти к выводам, стоит отметить, что обычно `OLS`-оценку выводят немного иначе, а именно минимизируя в явном виде длину вектора ошибок по коэффициентам $\hat{w}$, вернее, даже квадрат длины для удобства вычисления производных.
>
>$||\overrightarrow{e}||\rightarrow min$
>
>$||\overrightarrow{e}||^2\rightarrow min$
>
>$||\overrightarrow{b}-A\overrightarrow{w}||^2\rightarrow min$
>
>Формула получится точно такой же, какая есть у нас, просто способ вычислений будет не геометрический, а аналитический. Мы вернёмся к этому способу, когда будем обсуждать оптимизацию функции многих переменных в разделе по математическому анализу.

Наконец, мы может подвести итоги для случая «Всё сложно».

***

**Резюмируем ↓**

Если ранг матрицы $A$ меньше ранга расширенной системы $(A|b)$, то независимых уравнений больше, чем переменных $(rkA<(A|b)<m)$, а значит некоторые из них будут противоречить друг другу, то есть решений у системы нет.

Говоря на языке линейной алгебры, вектор $b$ линейно независим со столбцами матрицы $A$, а значит его нельзя выразить в качестве их линейной комбинации.

Однако можно получить приближённые решения по методу наименьших квадратов (`OLS` - оценка - $\hat{b}=(A^TA)^{-1}\cdot A^Tb$), идеей которого является ортогональная проекция вектора $b$ на столбцы матрицы $A$.

Найдите `OLS`-оценку для коэффициентов $w_1$, $w_2$ СЛАУ:

$\left\{\begin{array}{c} w_1+2w_2=1 \\ -3w_1+w_2=4 \\ w_1+2w_2=5 \\ w_1-w_2=0\end{array}\right.$

In [34]:
import numpy as np
A = np.array([[1,2],[-3,1],[1,2],[1,-1]])
b = np.array([1,4,5,0])
b = np.reshape(b,(4,1))


1. Вычислите матрицу Грама столбцов $A:A^{T}A=$

In [40]:
A.T@A

array([[12,  0],
       [ 0, 10]])

2. Вычислите матрицу $(A^{T}A)^{-1}$. Она имеет вид:

In [43]:
a_a = np.linalg.inv(A.T@A)
a_a

array([[0.08333333, 0.        ],
       [0.        , 0.1       ]])

3. Вычислите $A^{T} \overrightarrow{b}$

In [44]:
a_b = A.T@b
a_b

array([[-6],
       [16]])

4. Вычислите вектор оценок коэффициентов $\overrightarrow{w}$.

In [45]:
a_a@a_b

array([[-0.5],
       [ 1.6]])

# 3. Линейная регрессия МНК <a class="anchor" id=2></a>

[к содержанию](#0)

В задаче регрессии обычно есть **целевая переменная**, которую мы хотим предсказать. Её, как правило, обозначают буквой $y$. Помимо целевой переменной, есть **признаки** (их также называют **факторами** или **регрессорами**). Пусть их будет $k$ штук:

$y$ — целевая переменная

$x_1, x_2, \dots, x_k$ — признаки/факторы/регрессоры

Поставить задачу — значит ответить на два вопроса:

1. Что у нас есть?

2. Что мы хотим получить?

В задаче регрессии есть $N$ (как правило, их действительно много) наблюдений. Это наша обучающая выборка или датасет, представленный в виде таблицы. В столбцах таблицы располагаются векторы признаков $\overrightarrow{x_i}$.

$\overrightarrow{y} \in \mathbb{R}^N$

$x_1, x_2, \dots, x_k\in \mathbb{R}^N$

$\left(\begin{array}{} y_1 \\ y_2 \\ \dots \\ y_N\end{array}\right), \left(\begin{array}{} x_1 \\ x_2 \\ \dots \\ x_N\end{array}\right), \dots, \left(\begin{array}{} x_k1 \\ x_k2 \\ \dots \\ x_kN\end{array}\right)$

То есть и целевая переменная, и признаки представлены векторами из векторного пространства $\mathbb{R}^N$ — каждого вектора $N$ координат.

В качестве регрессионной модели мы будем использовать **модель линейной регрессии**. Мы предполагаем, что связь между целевой переменной и признаками линейная. Это означает, что:

$y=w_0+w_1x_1+w_2x_2+…+w_kx_k,$

или 

$y=(\overrightarrow{w}, \overrightarrow{x})$

Здесь $\overrightarrow{w}=(w_0,w_1,…,w_k)^T$ обозначают веса (коэффициенты уравнения линейной регрессии), а $\overrightarrow{x}=(1,x_1, x_2,…, x_k)^T$.

>Наличие коэффициента $w_0$ говорит о том, что мы строим регрессию с константой, или, как ещё иногда говорят, с **интерсептом**.

Пока что коэффициенты $w$ нам неизвестны. Как же их найти?

Для этого у нас есть $N$ наблюдений — обучающий набор данных.

Давайте попробуем подобрать такие веса $w$, чтобы для каждого наблюдения наше равенство было выполнено. Таким образом, получается $N$ уравнений на $k+1$ неизвестную.

$\left(\begin{array}{} y_1 \\ y_2 \\ \dots \\ y_N\end{array}\right)=w_0\cdot \left(\begin{array}{} 1 \\ 2 \\ \dots \\ 1 \end{array}\right)+w_1 \cdot\left(\begin{array}{} x_11 \\ x_12 \\ \dots \\ x_1N\end{array}\right)+ \dots +w_k\cdot \left(\begin{array}{} x_k1 \\ x_k2 \\ \dots \\ x_kN\end{array}\right)$

Или в привычном виде систем уравнений:

$\left\{\begin{array}{} w_0 1+w_1 x_{11} + \dots +w_k x_{k1}=y_1 \\ w_0 1+w_1 x_{12} + \dots +w_k x_{k2}=y_2 \\ \dots \\ w_0 1+w_1 x_{1N} + \dots +w_k x_{kN}=y_N \end{array}\right.$

>Говоря на языке машинного обучения, мы хотим обучить такую модель, которая описывала бы зависимость целевой переменной от факторов на обучающей выборке.

Как правило, $N$ гораздо больше $k$ (количество строк с данными в таблице намного больше количества столбцов) и система переопределена, значит точного решения нет. Поэтому можно найти только приближённое.

>**Примечание**. Полученной СЛАУ можно дать геометрическую интерпретацию. Если представить каждое наблюдение в виде точки на графике (см. ниже), то уравнение линейной регрессии будет задавать прямую (если фактор один) или гиперплоскость (если факторов $k$ штук). Приравняв уравнение прямой к целевому признаку, мы потребовали, чтобы эта прямая проходила через все точки в нашем наборе данных. Конечно же, это условие не может быть выполнено полностью, так как в данных всегда присутствует какой-то шум, и идеальной прямой (гиперплоскости) не получится, но зато можно построить приближённое решение.
>
><img src=m2_img2.png width=400>
>
>**Обратите внимание**, что у нас появился новый вектор из единиц. Он здесь из-за того, что мы взяли модель с интерсептом. Можно считать, что это новый регрессор-константа. Данная константа тянется из уравнения прямой, которое мы разбирали в модуле «ML-2. Обучение с учителем: регрессия».

Мы уже умеем решать переопределённые системы, для этого мы должны составить матрицу системы $A$, записав в столбцы все наши регрессоры, включая регрессор константу:

$A=\left(\begin{array}{} 1 & x_{11} & \dots & x_{k1} \\ 1 & x_{12} & \dots & x_{k2} \\ \dots & \dots & \dots & \dots \\ 1 & x_{1N} & \cdot & x_{kN} \end{array}\right) \Rightarrow N - строк, k+1 - столбец$

>**Примечание**. В контексте задач машинного обучения матрица $A$ называется **матрицей наблюдений**: по строкам отложены наблюдения (объекты), а по столбцам — характеризующие их признаки. В модулях по машинному обучению мы в основном обозначали её за $X$. Здесь же мы будем придерживаться традиций линейной алгебры и обозначать матрицу за $A$.
>
>**Примечание**. Обратите внимание, что индексация матрицы $A$ отличается от привычной нам индексации матрицы. Например, здесь $x_{12}$ — второе наблюдение первого регрессора. Это чистая формальность. Если обозначать за первый индекс номер наблюдения, а за второй индекс — номер регрессора, мы получим привычную нам нумерацию элементов матрицы (строка-столбец).

Осталось записать финальную формулу `OLS`-оценки для коэффициентов:

$\hat{\overrightarrow{w}} = (A^T A)^{-1} A^T \overrightarrow{y}$

Казалось бы, задача решена, однако это совсем не так, ведь мы искали коэффициенты не просто так, а чтобы сделать прогноз — предсказание на новых данных.

Допустим, у нас есть новое наблюдение по регрессорам, которое характеризуется признаками $\overrightarrow{x}_{NEW} = (x_{1, NEW}, x_{2, NEW}, ..., x_{k, NEW})^T$. Тогда, предсказание будет строиться следующим образом:

$\vec{y}_{NEW} = \vec{w}_0 + \vec{w}_1 x_{1, NEW} + ... + \vec{w}_k x_{k, NEW}$

или

$\vec{y}_{NEW} = (\hat{\vec{w}}, \vec{x}_{NEW})$

Теперь перейдём от формул к практике и решим задачу в контексте.

*** 
Рассмотрим классический датасет для обучения линейной регрессии — `Boston Housing`. В нём собраны усреднённые данные по стоимости недвижимости в 506 районах Бостона. Ниже вы видите фрагмент датасета.

Целевой переменной будет $\textcolor{red}{PRICE}$ — это, в некотором смысле, типичная (медианная) стоимость дома в районе.

Для примера возьмём в качестве регрессоров уровень преступности $\textcolor{blue}{CRIM}$ и среднее количество комнат в доме $\textcolor{blue}{RM}$.

<img src=m2_img3.png width=600>

Запишем нашу модель:

$y=w_0+w_1 \cdot x_1+w_2 \cdot x_2$

Для наглядности обозначим:

$y=w_0+w_1 \cdot CRIM+w_2 \cdot RM$

Составим матрицу регрессоров:

$A=\left(\begin{array}{} 1 & CRIM_1 & RM_1 \\ 1 & CRIM_2 & RM_2 \\ \dots & \dots & \dots \\ 1 & CRIM_N & RM_N \end{array}\right)$

В нашем случае $N=506$, а $k=2$. Размерность матрицы $A$ будет равна $\mathbb{dim}A=(506,3)$. Далее мы применяем формулу для вычисления оценок коэффициентов:

$\hat{\vec{w}} = (A^T A)^{-1} A^T \vec{y}$

Вычисления к этой задаче мы сделаем в `Python` ниже, а пока приведём конечный результат. Если сократить запись до двух знаков после точки, получим следующие коэффициенты:

$\hat{\vec{w}} = (-29.3, \ -0.26, \ 8.4)^T$

То есть:

$\hat{w}_0 = -29.3$

$\hat{w}_1 = -0.26$

$\hat{w}_2 = 8.4$

Мы можем переписать нашу модель для прогноза:

$\hat{y} = -29.3 - 0.26 \cdot CRIM + 8.4 \cdot RM$

Теперь, если у нас появятся новые наблюдения, то есть ещё один небольшой район с уровнем преступности 0.1 на душу населения и средним количеством комнат в доме, равным 8, мы сможем сделать прогноз на типичную стоимость дома в этом районе — 37 тысяч долларов:

$CRIM_{NEW} = 0.1$

$RM_{NEW} = 8$

$\hat{y}_{NEW} = -29.3 -0.26 \cdot 0.1 + 8.4 \cdot 8 \approx 37$

*** 
## → Решение на Python

In [46]:
# Загрузка библиотек
import numpy as np # для работы с массивами
import pandas as pd # для работы с DataFrame 
from sklearn import datasets # для импорта данных
import seaborn as sns # для визуализации статистических данных
import matplotlib.pyplot as plt # для построения графиков

# загружаем датасет
boston = datasets.load_boston()
boston_data = pd.DataFrame(
    data=boston.data, #данные
    columns=boston.feature_names #наименования столбцов
)
boston_data['PRICE'] = boston.target
boston_data.head()

c:\Users\AubakirovMA\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dat

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  PRICE  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

Формируем матрицу $A$ из столбца единиц и факторов $CRIM$ и $RM$, а также вектор целевой переменной $y$:

In [47]:
# составляем матрицу А и вектор целевой переменной
CRIM = boston_data['CRIM']
RM = boston_data['RM']
A = np.column_stack((np.ones(506), CRIM, RM))
y = boston_data[['PRICE']]
print(A)

[[1.0000e+00 6.3200e-03 6.5750e+00]
 [1.0000e+00 2.7310e-02 6.4210e+00]
 [1.0000e+00 2.7290e-02 7.1850e+00]
 ...
 [1.0000e+00 6.0760e-02 6.9760e+00]
 [1.0000e+00 1.0959e-01 6.7940e+00]
 [1.0000e+00 4.7410e-02 6.0300e+00]]


Посмотрим на размерность матрицы $A$:

In [48]:
# проверим размерность
print(A.shape)

(506, 3)


Теперь нам ничего не мешает вычислить оценку вектора коэффициентов $w$ по выведенной нами формуле МНК:

In [49]:
# вычислим OLS-оценку для коэффициентов
w_hat = np.linalg.inv(A.T@A)@A.T@y
print(w_hat.values)

[[-29.24471945]
 [ -0.26491325]
 [  8.39106825]]


Теперь составим прогноз нашей модели:

In [50]:
# добавились новые данные:
CRIM_new = 0.1
RM_new = 8
# делаем прогноз типичной стоимости дома
PRICE_new = w_hat.iloc[0]+w_hat.iloc[1]*CRIM_new+w_hat.iloc[2]*RM_new
print(PRICE_new.values)

[37.85733519]


Согласитесь, такая запись вычисления оценки стоимости слишком длинная и неудобная, особенно если факторов не два, как у нас, а 200. Более короткий способ сделать прогноз — вычислить скалярное произведение вектора признаков и коэффициентов регрессии.

Для удобства дальнейшего использования оформим характеристики нового наблюдения в виде матрицы размером $(1, 3)$:

In [51]:
new=np.array([[1,CRIM_new,RM_new]])
print('prediction:', (new@w_hat).values)

prediction: [[37.85733519]]


>**Примечание**. Обратите внимание, что, решая задачу с помощью `Python`, мы получили немного другой результат прогноза стоимости. Это связано с тем, что при выполнении ручного расчёта мы округлили значения коэффициентов и получили менее точный результат.

Мы уже знаем, что алгоритм построения модели линейной регрессии по МНК реализован в классе `LinearRegression`, находящемся в модуле `sklearn.linear_model`. Для вычисления коэффициентов (обучения модели) нам достаточно передать в метод `fit()` нашу матрицу с наблюдениями и вектор целевой переменной, а для построения прогноза — вызвать метод `predict()`:

In [52]:
from sklearn.linear_model import LinearRegression
# создаём модель линейной регрессии
model = LinearRegression(fit_intercept=False)
# вычисляем коэффициенты регрессии
model.fit(A, y)
print('w_hat:', model.coef_)
new_prediction = model.predict(new)
print('prediction:', new_prediction)

w_hat: [[-29.24471945  -0.26491325   8.39106825]]
prediction: [[37.85733519]]


>Примечание. Здесь при создании объекта класса `LinearRegression` мы указали `fit_itercept=False`, так как в нашей матрице наблюдений $A$ уже присутствует столбец с единицами для умножения на свободный член $w_0$. Его повторное добавление не имеет смысла.

Сделайте прогноз типичной стоимости (в тыс. долларов) дома в городе с уровнем преступности $CRIM=0.2$ и средним количеством комнат в доме $RM=6$. В качестве модели используйте линейную регрессию, оценка вектора коэффициентов которой равна: $\hat{\vec{w}} = (-29.3, \ -0.26, \ 8.4)$

In [53]:
# добавились новые данные:
CRIM_new = 0.2
RM_new = 6
# делаем прогноз типичной стоимости дома
PRICE_new = w_hat.iloc[0]+w_hat.iloc[1]*CRIM_new+w_hat.iloc[2]*RM_new
print(PRICE_new.values)

[21.04870738]


## ПРОБЛЕМЫ В КЛАССИЧЕСКОЙ МНК-МОДЕЛИ

Заметим, что в уравнении классической `OLS`-регрессии присутствует очень важный множитель $A^TA$:

$\vec{w}=\left(A^{T} A\right)^{-1} A^{T}\vec{y}$

Вы могли заметить, что это матрица Грама значений наших признаков, включая признак-константу.

Вспомним свойства этой матрицы: 

* квадратная (размерности $k+1$ на $K+1$, где $k$ — количество факторов);

* симметричная.

>Как и у любого метода, у классической `OLS`-регрессии есть свои **ограничения**. Если матрица $A^TA$ вырождена или близка к вырожденной, то хорошего решения у классической модели не получится. Такие данные называют **плохо обусловленными**.

***

Корректна ли модель классической `OLS`-регрессии, если

$\vec{y}=\left(\begin{array}{} 1\\2\\5\\1\end{array}\right), \vec{x_1}=\left(\begin{array}{} 2\\1\\1\\2\end{array}\right), \vec{x_2}=\left(\begin{array}{} -2\\-1\\-1\\-2\end{array}\right)$

Запишем матрицу $A$ и вычислим $A^TA$:

$A=(\vec{1}, \vec{x_1}, \vec{x_2})=\left(\begin{array}{} 1&2&-2\\1&1&-1\\1&1&-1\\1&2&-2\end{array}\right)$

$A^TA=\left(\begin{array}{} 1&1&1&1\\2&1&1&2\\-2&-1&-1&-2\end{array}\right)\cdot\left(\begin{array}{} 1&2&-2\\1&1&-1\\1&1&-1\\1&2&-2\end{array}\right)=\left(\begin{array}{} 4&6&-6\\6&10&-10\\-6&-10&10\end{array}\right)$

Как видите, две последние строки матрицы $A^TA$ являются пропорциональными. Это говорит о том, что матрица вырождена $(det A^T A =0)$ или её ранг $(rkA)$ меньше количества неизвестных $(3)$, а значит обратной матрицы $(A^TA)^{-1}$ к ней не существует. Отсюда следует, что классическая `OLS`-модель **неприменима для этих данных**.

>Борьба с вырожденностью матрицы $A^TA$ часто сводится к устранению «плохих» (зависимых) признаков. Для этого анализируют корреляционную матрицу признаков или матрицу их значений. Но иногда проблема может заключаться, например, в том, что один признак измерен в тысячных долях, а другой — в тысячах единиц. Тогда коэффициенты при них могут отличаться в миллион раз, что потенциально может привести к вырожденности матрицы $A^TA$.

В устранении этой проблемы может помочь знакомая нам **нормализация/стандартизация данных**.

## ОСОБЕННОСТИ КЛАССА LINEAR REGRESSION БИБЛИОТЕКИ SKLEARN

Давайте посмотрим, что «скажет» `Python`, если мы попробуем построить модель линейной регрессии на вырожденной матрице наблюдений, используя классическую формулу линейной регрессии.

In [54]:
# создадим вырожденную матрицу А
A = np.array([
    [1, 1, 1, 1], 
    [2, 1, 1, 2], 
    [-2, -1, -1, -2]]
).T
y = np.array([1, 2, 5, 1])
# вычислим OLS-оценку для коэффициентов
w_hat=np.linalg.inv(A.T@A)@A.T@y
print(w_hat) 

LinAlgError: Singular matrix

Как и ожидалось, мы получили ошибку, говорящую о том, что матрица $A^TA$ — сингулярная (вырожденная), а значит обратить её не получится. Что и требовалось доказать — с математикой всё сходится.

⭐ Настало время фокусов!

Попробуем обучить модель линейной регрессии `LinearRegression` из модуля `sklearn`, используя нашу вырожденную матрицу :

In [57]:
# создаём модель линейной регрессии
model = LinearRegression(fit_intercept=False)
# вычисляем коэффициенты регрессии
model.fit(A, y)
print('w_hat:', model.coef_)

w_hat: [ 6.   -1.25  1.25]


Никакой ошибки не возникло! Более того, у нас даже получились вполне адекватные оценки коэффициентов линейной регрессии $\hat{\vec{w}}$.

?Но ведь мы только что использовали формулу для вычисления коэффициентов при расчётах вручную и получали ошибку. Как мы могли получить результат, если матрица $A^TA$ вырожденная? Существование обратной матрицы для неё противоречит законам линейной алгебры. Неужели это очередной случай, когда «мнения» математики и Python расходятся?

На самом деле, не совсем. Здесь нет никакой магии, ошибки округления или бага. Просто в реализации линейной регрессии в `sklearn` предусмотрена **борьба с плохо определёнными (близкими к вырожденным и вырожденными) матрицами**.

>Для этого используется метод под названием **сингулярное разложение** (`SVD`). О нём мы будем говорить отдельно, однако уже сейчас отметим тот факт, что данный метод позволяет всегда получать корректные значения при обращении матриц.
>
>Если вы хотите понять, почему так происходит, ознакомьтесь с этой [статьёй](https://towardsdatascience.com/understanding-linear-regression-using-the-singular-value-decomposition-1f37fb10dd33).
>
>**Суть метода** заключается в том, что в `OLS`-формуле мы на самом деле используем не саму матрицу $A$, а её диагональное представление из сингулярного разложения, которое гарантированно является невырожденным. Вот и весь секрет.

?Правда, открытым остаётся вопрос: **можно ли доверять коэффициентам**, полученным таким способом, и интерпретировать их? 

В дальнейшем мы увидим, что делать этого лучше не стоит: возможна такая ситуация, при которой коэффициенты при линейно зависимых факторах, которые получаются в результате применения линейной регрессии через сингулярное разложение, могут получиться слишком большими по модулю. Они могут измеряться миллионами, миллиардами и более высокими порядками, что не будет иметь отношения к действительности. Такие коэффициенты не подлежат интерпретации.

Заметим, что в случае использования решения через сингулярное разложение для линейно зависимых столбцов коэффициенты будут всегда получаться одинаковыми по модулю, но различными по знаку: $w_1=-1.25$ и $w_2=1.25$. Неудивительно, ведь второй и третий столбцы матрицы $A$ линейно зависимы с коэффициентом $-1$.

Запишем итоговое уравнение линейной регрессии:

$y=w_{0} \overrightarrow{1}+w_{1} \vec{x}_{1}+w_{2} \vec{x}_{2}=6-1.25 \cdot \vec{x}_{1}+1.25 \cdot \vec{x}_{2},$

поставим столбцы матрицы $A$ в данное уравнение, чтобы получить прогноз:

$\vec{y}=6\left(\begin{array}{} 1\\1\\1\\1\end{array}\right)-1.25\left(\begin{array}{} 2\\1\\1\\2\end{array}\right)+1.25\left(\begin{array}{} -2\\-1\\-1\\-2\end{array}\right)=\left(\begin{array}{} 1\\3.5\\3.5\\1\end{array}\right)$

>**Примечание**. На самом деле сингулярное разложение зашито в функцию `np.linalg.lstsq()`, которая позволяет в одну строку построить модель линейной регрессии по МНК:
>
>классическая `OLS`-регрессия в numpy с возможностью получения решения даже для вырожденных матриц
>
>`np.linalg.lstsq(A, y, rcond=None)`
>
>Функция возвращает четыре значения:
>
>* вектор рассчитанных коэффициентов линейной регрессии;
>
>* сумму квадратов ошибок, `MSE` (она не считается, если ранг матрицы $A$ меньше числа неизвестных, как в нашем случае);
>
>* ранг матрицы $A$;
>
>* вектор из сингулярных значений, которые как раз и оберегают нас от ошибки (о них мы поговорим позже).
>
>Обратите внимание, что мы получили те же коэффициенты, что и с помощью `sklearn`. При этом ранг матрицы $A$ равен 2, что меньше количества неизвестных коэффициентов. Это ожидаемо говорит о вырожденности матрицы $A$ и, как следствие, матрицы $A^TA$.

*** 
## Резюмируем ↓

* Для поиска коэффициентов модели линейной регрессии используется МНК-оценка: 

$\vec{w}=\left(A^{T} A\right)^{-1} A^{T} \vec{y}$

* Полученная матричная формула не зависит от размерности матрицы наблюдений $A$ и работает при любом количестве объектов/признаков в данных.

* Для реализации обучения модели линейной регрессии по `МНК` в `sklearn` используется класс `LinearRegression`.

* Для предотвращения обращения вырожденной матрицы $A$ в `LinearRegression` вместо самой матрицы используется её сингулярное разложение. Поэтому на практике при построении модели линейной регрессии вместо ручного вычисления обратной матрицы с помощью `np.inv()` приоритетнее пользоваться именно `LinearRegression` из `sklearn` (или `np.linalg.lstsq()`).

Данный метод оберегает от ошибки только при обращении плохо обусловленных и вырожденных матриц и не гарантирует получение корректных коэффициентов линейной регрессии.

# 4. Стандартизация векторов и матрица корреляции <a class="anchor" id=2></a>

[к содержанию](#0)

## СТАНДАРТИЗАЦИЯ ВЕКТОРОВ

В модулях по разведывательному анализу данных и машинному обучению мы не раз говорили о преобразованиях признаков путём нормализации и стандартизации. Вспомним, что это такое ↓

>**Нормализация** — это процесс приведения признаков к единому масштабу, например от 0 до 1. Пример — min-max-нормализация:
>
>$x_{scaled} =  \frac{x - x_{min}}{x_{max} - x_{min}}$
>
>**Стандартизация** — это процесс приведения признаков к единому масштабу характеристик распределения — нулевому среднему и единичному стандартному отклонению:
>
>$x_{scaled} =  \frac{x - x_{mean}}{x_{std}}$

В линейной алгебре под стандартизацией вектора $\vec{x} \in \mathbb{R}^n$ понимается несколько другая операция, которая проходит в два этапа:

1. Центрирование вектора — это операция приведения среднего к 0:

$\vec{x}_{cent} = \vec{x} - \vec{x}_{mean}$

2. Нормирование вектора — это операция приведения диапазона вектора к масштабу от -1 до 1 путём деления центрированного вектора на его длину:

$\vec{x}_{st} =  \frac{\vec{x}_{cent}}{ \| \vec{x}_{cent} \| },$

где $\vec{x}_{mean}$ — вектор, составленный из среднего значения вектора $\vec{x}$, а $\| \vec{x}_{cent} \|$ — длина вектора  $\vec{x}_{cent}$.

В результате стандартизации вектора всегда получается новый вектор, длина которого равна 1:

$\| \vec{x}_{st} \|  = 1$

**Пример № 1**

Необходимо стандартизировать векторы:

$\vec{x_1}=\left(\begin{array}{} 1\\2\\3 \end{array}\right)$ и $\vec{x_2}=\left(\begin{array}{} 3000\\1000\\2000 \end{array}\right)$

Центрируем:

$\vec{x_1}=\frac{1+2+3}{3}=3$

$\vec{x}_{1cent}=\left(\begin{array}{} 1\\2\\3 \end{array}\right)-\left(\begin{array}{} 3\\3\\3 \end{array}\right)=\left(\begin{array}{} -2\\-1\\3 \end{array}\right)$

$\vec{x_2}=\frac{3000+1000+2000}{3}=2000$

$\vec{x}_{2cent}=\left(\begin{array}{} 3000\\1000\\2000 \end{array}\right)-\left(\begin{array}{} 2000\\2000\\2000 \end{array}\right)=\left(\begin{array}{} 1000\\-1000\\0 \end{array}\right)$

Нормируем:

$\| \vec{x}_{1cent} \| =\sqrt{(-2)^2 + (-1)^2 + 3^2}=\sqrt{14}$

$\| \vec{x}_{1st} \| = \frac{1}{\| \vec{x}_{1cent} \|}\cdot\vec{x}_{1cent}=\frac{1}{\sqrt{14}}\cdot\left(\begin{array}{} -2\\-1\\3 \end{array}\right)\approx\left(\begin{array}{} -0.535\\-0.267\\0.802 \end{array}\right)$

$. $

$\| \vec{x}_{2cent} \| =\sqrt{(1000)^2 + (-1000)^2 + 0^2}=1000\sqrt{2}$

$\| \vec{x}_{2st} \| = \frac{1}{\| \vec{x}_{2cent} \|}\cdot\vec{x}_{2cent}=\frac{1}{1000\sqrt{2}}\cdot\left(\begin{array}{} 1000\\-1000\\0 \end{array}\right)\approx\left(\begin{array}{} 0.707\\-0.707\\0 \end{array}\right)$

Как видите, теперь оба признака имеют значения от -1 до 1 и равный порядок, в отличие от исходных признаков.

Давайте посмотрим, что произойдёт с матрицей Грама после стандартизации векторов $\vec{x}_1$ и $\vec{x}_2$:

**Пример № 2**

Найти матрицу для стандартизированных признаков для

$\vec{x_1}=\left(\begin{array}{} 1\\2\\3 \end{array}\right)$ и $\vec{x_2}=\left(\begin{array}{} 3000\\1000\\2000 \end{array}\right)$

Вычислим попарные скалярные произведения новых признаков:

$G(\vec{x}_{1,st},\vec{x}_{2,st})=G(\left(\begin{array}{} -0.535\\-0.267\\0.802 \end{array}\right),\left(\begin{array}{} 0.707\\-0.707\\0 \end{array}\right))=\left(\begin{array}{} (\vec{x}_{1,st},\vec{x}_{1,st}) & (\vec{x}_{1,st},\vec{x}_{2,st}) \\ (\vec{x}_{2,st},\vec{x}_{1,st}) & (\vec{x}_{2,st},\vec{x}_{2,st})  \end{array}\right)=\left(\begin{array}{} 1 & -0.189 \\ -0.189 & 1 \end{array}\right)$

Как видите, все числа — в диапазоне от -1 до 1. 

>Забегая вперёд, скажем, что это так называемые **выборочные корреляции признаков**, а сама матрица является **матрицей корреляций** или **корреляционной матрицей**. Пока просто запомните, как выглядит эта матрица.

Вот **ещё одна особенность стандартизации** ↓

До стандартизации мы прогоняли регрессию $y$ на регрессоры $x_1, x_2, …, x_k$ и константу. Всего получалось $k+1$ коэффициентов:

$\vec{y}=w_{0}+w_{1} \vec{x}_{1}+w_{2} \vec{x}_{2}+\ldots+w_{k} \vec{x}_{k}$

После стандартизации мы прогоняем регрессию стандартизованного $y$ на стандартизованные регрессоры **без константы**:

$\vec{y}=w_{1_{st}} \vec{x}_{1_{st}}+w_{2_{st}} \vec{x}_{2_{st}}+\ldots+w_{k_{st}} \vec{x}_{k_{st}}$

Математически мы получим одну и ту же регрессию в том смысле, что если пересчитать стандартизированные коэффициенты, мы получим исходные. То же и с прогнозом (пересчёт здесь опустим).

***

**В ЧЁМ БОНУСЫ?**

Математика говорит, что регрессия исходного $y$ на исходные («сырые») признаки c константой точно такая же, как регрессия стандартизированного на стандартизированные признаки без константы. В чём же разница? Математически — ни в чём.

На прогноз модели линейной регрессии, построенной по МНК, и её качество стандартизация практически не влияет. Масштабы признаков будут иметь значение только в том случае, если для поиска коэффициентов вы используете численные методы, такие как градиентный спуск (`SGDRegressor` из `sklearn`). О нём мы поговорим, когда будем знакомиться с алгоритмом градиентного спуска в модуле по оптимизации.

Однако с точки зрения интерпретации важности коэффициентов разница есть. Если вы занимаетесь отбором наиболее важных признаков по значению коэффициентов линейной регрессии на нестандартизированных данных, это будет не совсем корректно: один признак может изменяться от 0 до 1, а второй — от -1000 до 1000. Коэффициенты при них также будут различного масштаба. Если же вы посмотрите оценки коэффициентов регрессии после стандартизации, то они будут в едином масштабе, что даст более цельную и объективную картину.

Более важный бонус заключается в том, что после стандартизации матрица Грама признаков как по волшебству превращается в корреляционную матрицу, о которой пойдёт речь далее. Почему это хорошо? На свойства корреляционной матрицы опираются такие алгоритмы, как метод главных компонент и сингулярное разложение, а так как «сырая» и стандартизированная регрессия математически эквивалентны, то имеет смысл исследовать стандартизированную, а результаты обобщить на «сырую».

**Пример № 3**

Вновь рассмотрим данные о стоимости жилья в районах Бостона.

На этот раз возьмём четыре признака: `CHAS`, `LSTAT`, `CRIM` и `RM`.

In [58]:
boston_data[['CHAS', 'LSTAT', 'CRIM','RM']].describe()

CHAS       LSTAT        CRIM          RM
count  506.000000  506.000000  506.000000  506.000000
mean     0.069170   12.653063    3.613524    6.284634
std      0.253994    7.141062    8.601545    0.702617
min      0.000000    1.730000    0.006320    3.561000
25%      0.000000    6.950000    0.082045    5.885500
50%      0.000000   11.360000    0.256510    6.208500
75%      0.000000   16.955000    3.677083    6.623500
max      1.000000   37.970000   88.976200    8.780000

Видим, что каждый из признаков измеряется в различных единицах и изменяется в различных диапазонах: например, `CHAS` лежит в диапазоне от 0 до 1, а вот `CRIM` — в диапазоне от 0.006 до 88.976.

Рассмотрим модель линейной регрессии по МНК без стандартизации. Помним, что необходимо добавить столбец из единиц:

In [59]:
# составляем матрицу наблюдений и вектор целевой переменной
A = np.column_stack((np.ones(506), boston_data[['CHAS', 'LSTAT', 'CRIM','RM']]))
y = boston_data[['PRICE']]
# вычисляем OLS-оценку для коэффициентов без стандартизации
w_hat=np.linalg.inv(A.T@A)@A.T@y
print(w_hat.values)

[[-1.92052548]
 [ 3.9975594 ]
 [-0.58240212]
 [-0.09739445]
 [ 5.07554248]]


Вот наши коэффициенты. Округлим их для наглядности:

$\hat{w}_0 = -1.92$

$\hat{w}_{CHAS} = 4$

$\hat{w}_{LSTAT} = -0.6$

$\hat{w}_{CRIM} = -0.1$

$\hat{w}_{RM} = 5$

Давайте вспомним интерпретацию коэффициентов построенной модели линейной регрессии, которую мы изучали в модуле «ML-2. Обучение с учителем: регрессия». Значение коэффициента $\hat{w}_i$ означает, на сколько в среднем изменится медианная цена (в тысячах долларов) при увеличении $x_i$ на 1.

Например, если количество низкостатусного населения (`LSTAT`) увеличится на 1 %, то медианная цена домов в районе (в среднем) упадёт на 0.1 тысяч долларов. А если среднее количество комнат (`RM`) в районе станет больше на 1, то медианная стоимость домов в районе (в среднем) увеличится на 5 тысяч долларов. 

>Тут в голову может прийти мысль: судя по значению коэффициентов, количество комнат (`RM`) оказывает на стоимость жилья большее влияние, чем процент низкостатусного населения (`LSTAT`). Однако **такой вывод будет ошибочным**. Мы не учитываем, что признаки, а значит и коэффициенты линейной регрессии, лежат в разных масштабах. Чтобы говорить о важности влияния признаков на модель, нужно строить её на стандартизированных данных.

Помним, что для построения стандартизированной линейной регрессии нам не нужен вектор свободных коэффициентов, а значит и столбец из единиц тоже не понадобится.

Сначала центрируем векторы, которые находятся в столбцах матрицы $A$. Для этого вычтем среднее, вычисленное по строкам матрицы $A$ в каждом столбце, с помощью метода `mean()`. Затем разделим результат на длины центрированных векторов, вычисленных с помощью функции `linalg.norm()`.

>**Примечание**. Обратите внимание, что для функции `linalg.norm()` обязательно необходимо указать параметр `axis=0`, так как по умолчанию норма считается для всей матрицы, а не для каждого столбца в отдельности. С определением нормы матрицы и тем, как она считается, вы можете ознакомиться в [документации к функции norm()](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html).


In [60]:
# составляем матрицу наблюдений без дополнительного столбца из единиц
A = boston_data[['CHAS', 'LSTAT', 'CRIM','RM']]
y = boston_data[['PRICE']]
# стандартизируем векторы в столбцах матрицы A
A_cent = A - A.mean()
A_st = A_cent/np.linalg.norm(A_cent, axis=0)
A_st.describe().round(2)

CHAS   LSTAT    CRIM      RM
count  506.00  506.00  506.00  506.00
mean    -0.00   -0.00    0.00   -0.00
std      0.04    0.04    0.04    0.04
min     -0.01   -0.07   -0.02   -0.17
25%     -0.01   -0.04   -0.02   -0.03
50%     -0.01   -0.01   -0.02   -0.00
75%     -0.01    0.03    0.00    0.02
max      0.16    0.16    0.44    0.16

Теперь векторы имеют одинаковые средние значения и стандартные отклонения. Если вычислить длину каждого из векторов, мы увидим, что они будут равны 1:

In [61]:
print(np.linalg.norm(A_st, axis=0))

[1. 1. 1. 1.]


Для получения стандартизированных коэффициентов нам также понадобится стандартизация целевой переменной $y$ по тому же принципу:

In [62]:
# стандартизируем вектор целевой переменной
y_cent = y - y.mean()
y_st = y_cent/np.linalg.norm(y_cent)

Формула для вычисления коэффициента та же, что и раньше, только матрица $A$ теперь заменяется на $A_{st}$, а $y$ — на $y_{st}$:

In [63]:
# вычислим OLS-оценку для стандартизированных коэффициентов
w_hat_st=np.linalg.inv(A_st.T@A_st)@A_st.T@y_st
print(w_hat_st.values)

[[ 0.11039956]
 [-0.45220423]
 [-0.09108766]
 [ 0.38774848]]


Вновь смотрим на коэффициенты. Помним, что коэффициента $\hat{w}_0$ у нас больше нет:

$\hat{w}_{CHAS} = 0.11$

$\hat{w}_{LSTAT} = -0.45$

$\hat{w}_{CRIM} = -0.09$

$\hat{w}_{RM} = 0.38$

Итак, мы видим картину, прямо противоположную той, что видели ранее. Теперь модуль коэффициента $\left|\hat{w}_{LSTAT, \ st} \right| = 0.45$ будет выше, чем модуль коэффициента $\left|\hat{w}_{RM, \ st} \right| = 0.38$. Значит, процент низкостатусного населения оказывает большее влияние на значение стоимости жилья, чем количество комнат.

Однако теперь интерпретировать сами коэффициенты в тех же измерениях у нас не получится.

***

**Сделаем важный вывод** ↓

Для того чтобы проинтерпретировать оценки коэффициентов линейной регрессии (понять, каков будет прирост целевой переменной при изменении фактора на 1 условную единицу), нам достаточно построить линейную регрессию в обычном виде без стандартизации и получить обычный вектор $\hat{\vec{w}}$.

Однако, чтобы корректно говорить о том, какой фактор оказывает на прогноз большее влияние, необходимо рассматривать стандартизированную оценку вектора коэффициентов $\hat{\vec{w}}_{st}$.

Давайте поближе взглянем на матрицу Грама для стандартизированных факторов:

In [64]:
# матрица Грама
A_st.T @ A_st

CHAS     LSTAT      CRIM        RM
CHAS   1.000000 -0.053929 -0.055892  0.091251
LSTAT -0.053929  1.000000  0.455621 -0.613808
CRIM  -0.055892  0.455621  1.000000 -0.219247
RM     0.091251 -0.613808 -0.219247  1.000000

На самом деле мы с вами только что вычислили **матрицу выборочных корреляций** наших исходных факторов. Мы уже сталкивались с ней много раз в разделах по разведывательному анализу данных и машинному обучению, правда, вычисляли её мы с помощью функции `Pandas`, а теперь научились делать это вручную.

>**Примечание**. Матрицу корреляций можно получить только в том случае, если производить стандартизацию признаков как векторы (делить на длину центрированного вектора $\vec{x}_{st}$). Другие способы стандартизации/нормализации признаков не превращают матрицу Грама в матрицу корреляций.

# 5. Практика. Лин.регрессия МНК <a class="anchor" id=2></a>

[к содержанию](#0)

# 6. Полиноминальная регрессия <a class="anchor" id=2></a>

[к содержанию](#0)

# 7. Регуляция <a class="anchor" id=2></a>

[к содержанию](#0)

# 8. Практика. Полиноминальная регрессия и регуляция <a class="anchor" id=2></a>

[к содержанию](#0)

# 9. Итоги <a class="anchor" id=2></a>

[к содержанию](#0)